### PR TITLE
Multi-character and multi-byte delimiter support & partial use of es6 (nodejs requirement latest 6.x branch)

### PR DESCRIPTION
**Pull request by ghost**
* Added test cases where the multi-byte and multi-char delimiters are in use
* Switched from string concatenation to a buffer for safer delimiting (especially the case for incomplete data-event emission)

### CODE REVIEW

**Comment by peterhaldbaek** [comment]
Sweet! Thanks for the contribution, I really like the Buffer part. I guess it makes it a lot faster as well? Not sure when I have the time to review it properly but hopefully it will be within the next week. 👍

**Comment by ghost** [comment]
The benefit of leaving it as a buffer means that the data is only transformed to a String when needed.

In the first commit I've implemented the buffer but relied on the String split.. it turned out to be a bad idea when the delimiter character was multi-byte or if it was a multi character string.

So String split would be out of question, I have switched to a different method where I iterate through the received buffer and try finding the delimiter buffer sequence inside the chunk and then split the messages inside the chunk to an array containing the buffers.

That way, in case I received an incomplete multi-byte delimiter at the end of a chunk, it would not be corrupted to a � and when the next chunk arrived would be properly completed for splitting.

**Comment by peterhaldbaek** [comment]
That makes sense. My usecase for creating this was much simpler as I was parsing ANSI characters and the delimiter was a single character. Your addition improves the stream a lot so I see no reason to reject it right now. If anything I would probably prefer a single commit but I can probably fix it up on my end.

**Comment by peterhaldbaek** [comment]
Is the package-lock.json file necessary? Been away from Node for a while but reading up on it it seems like it should not be added for libraries. See https://stackoverflow.com/q/44206782 (especially the comments).

File: delimiter-stream.js [hunk 1]
```diff
@@ -1,54 +1,103 @@
...
+  encoding: 'utf8'
+};
 
 util.inherits(DelimiterStream, Transform);
 
-function DelimiterStream(options) {
+function DelimiterStream (options) {
   Transform.call(this, options);
 
-  var defaults = {
```
Comment (peterhaldbaek) [inline delimiter-stream.js:10]: What is the reason for moving the defaults out of the class definition?

**Review by peterhaldbaek — COMMENTED** [review]
Looked at it more closely and it looks really good. Nice work!

File: delimiter-stream.js (hunk 1 above)
Comment (ghost) [reply delimiter-stream.js:10]: I moved it outside of the constructor as it's always the same information and didn't see why the defaults object has to be created every time a new DelimiterStream instance is made.

**Comment by ghost** [comment]
The package-lock.json ensures that the installed dependencies stay the same for development and deployment / whenever an "npm install" is executed.

That way there are no discrepancies between installations, it made sense to me

**Comment by peterhaldbaek** [comment]
Yes, I see the reasons for having it. I was just reading up on different opinions on the matter and started wondering whether it was needed or not. Since other projects depending on this will not be affected by the lock file I am okay with leaving it as it makes it easier to reproduce issues (as you argue). I will keep it.

**Comment by ghost** [comment]
An issue right now would be that the changes with the new language features would drop support for anything below Node 6.x

Node 4.x end of life will be reached by the end of the month. https://github.com/nodejs/Release

**Comment by peterhaldbaek** [comment]
Yes, I noticed that but don't think it is an issue. As you mention Node 4 is very close to EOL so no matter what it is time to move on. We could also make the version 2.0.0 as it could be considered a breaking change but it seems a little overkill.

**Pull request merged**